moneriote.py

```diff
@@ -1,303 +1,502 @@
+        :return: valid nodes
+        """
+        if len(nodes) == 0:
+            return nodes
+
+        now = datetime.now()
+        log_msg('Scanning %d node(s) on port %d. This can take several minutes. Let it run.' % (len(nodes), self._m_rpc_port))
+
+        pool = Pool(processes=self._concurrent_scans)
+        nodes = RpcNodeList.from_list(pool.map(partial(RpcNode.is_valid, self._blockchain_height), nodes))
+        pool.close()
+        pool.join()
+
+        log_msg('Scanning %d node(s) done after %d seconds, found %d valid' % (
+            len(nodes), (datetime.now() - now).total_seconds(), len(nodes.valid(valid=True))))
+
+        if remove_invalid:
+            nodes = nodes.valid(valid=True)
+        return nodes
+
+    def loop(self):
+        # get & set the current blockheight
+        height = self.monerod_get_height(method=self.md_height_discovery_method)
+        if not height or not isinstance(height, int):
+            log_err("Unable to fetch the current blockchain height")
+            return
+
+        self._blockchain_height = height
+
+        # get the correct zone
+        if not self.cf_zone_id:
+            log_msg('Determining zone_id; looking for \'%s\'' % self.cf_domain_name)
+            zones = self._cf_request(url='')
+            try:
+                self.cf_zone_id = next(zone.get('id') for zone in zones if zone.get('name') == self.cf_domain_name)
+            except StopIteration:
+                log_err('could not determine zone_id. Is your Cloudflare domain correct?')
+                sys.exit()
+            log_msg('Cloudflare zone_id \'%s\' matched to \'%s\'' % (self.cf_zone_id, self.cf_domain_name))
+
+        # fetch existing A records from Cloudflare
+        nodes_cf = RpcNodeList()
+
+        log_msg('Fetching existing record(s) (%s.%s)' % (self.cf_subdomain_name, self.cf_domain_name))
+        for record in self._cf_request('%s/dns_records/' % self.cf_zone_id):
+            # filter on A records / subdomain
+            if record.get('type') == 'A' and record.get('name') == self.cf_fulldomain_name:
+                node = RpcNode(address=record.get('content'),
+                               cf_id=record.get('id'),
+                               port=self._m_rpc_port)
+                nodes_cf.append(node)
+        log_msg("Found %d existing record(s) on Cloudflare" % len(nodes_cf))
+
+        # Validate existing Cloudflare records.
+        if len(nodes_cf):
+            nodes_cf = self.scan(nodes_cf)
+
+            # Remove invalid nodes, if at all.
+            for invalid_node in nodes_cf.valid_cf(valid=False):
+                self.cf_delete_record(invalid_node)
+            nodes_cf = nodes_cf.valid(valid=True)
+
+        # If we have room in Cloudflare for records, we need to discover nodes
+        cf_add_count = self.cf_max_records - len(nodes_cf)
+        if cf_add_count <= 0:
+            log_msg("No need to add more Cloudflare records")
```

Comment:
If I read this right, it seems your strategy is keeping those max_records until some of them go offline. Then script find new node to add. However, I prefer my previous way that keep all valid node IP in current_node (and file it). Then we random pick max_records numbers of IP and update to CF. This could spread the traffic and reduce the loading of volunteer's node.

Reply:
Also, I believe keeping all IP is easier to make geo-ip subdomain (my next task).

Reply:
I agree on the refreshing of existing records part. This however, would not require a client side file. 

On the topic of geo-ip subdomains; you also do not need a client-side file. It's just a simple ipv4->geo key/val lookup. There are Python libs for the Maxmind dump.

The only valid technical reason to use a cache file is to not scan peers multiple times in a short time span. This requirement would also dictate managing when peers have entered the cache file; at which point my interest go towards Redis.

I also have more plans with this repository, would you say I should hard-fork this and continue on my own?

Reply:
I would merge this nice PR if the refreshing of existing record part could be done. But if you have more plans and think hard fork is better for your work then I'll totally OK with that.

The original reason I kept all IP in file is when switching to a new synced node, the peer list would need some time to complete. So keep IP list is more efficient. This often happens when I test this script in early time. But now I agree it might not necessary. And my maintained geo-ip is the point of keeping all IP, not file, so you're right.

Reply:
Ok I will see what I can do.